### PR TITLE
Many small AArch64 fixes

### PR DESCRIPTION
**Pull request by theRealAph**
These fix a bunch of minor problems in the AArch64 back end. They are impossible to test in isolation; they are the minimum needed to run a any tests.

They are:

Tableswitch completely rewritten. It generated incorrect code and did not properly construct a jump table.

Narrow OOPs were not supported at all. All done.

All memory loads sign extended. Rewritten to use either zero- or sign-extension depending on context.

Sign and zero extension rewritten in many places.

Bit scan operations returned an incorrect type, causing compiler errors.

CompareAndSwap returned a value instead of a boolean.

ByteSwap was unimplemented.

PatchReturnAddressOp did not work at all.

AArch64HotSpotSafepointOp did not use the correct relocation.

Temporary register usage in stack spills was incorrect. Stack-stack spills were not implemented at all.

Loading a stack address corrupted the stack pointer.

### CODE REVIEW

File: graal/com.oracle.graal.asm.aarch64/src/com/oracle/graal/asm/aarch64/AArch64MacroAssembler.java [hunk 1]
```diff
@@ -596,6 +624,8 @@ public void neg(int size, Register dst, Register src) {
      */
     @Override
     public void add(int size, Register dst, Register src, int immediate) {
+        if ((immediate & 7) != 0)
```
Comment (rschatz) [inline graal/com.oracle.graal.asm.aarch64/src/com/oracle/graal/asm/aarch64/AArch64MacroAssembler.java]: Please always use `{ }` even for one-liners.

**Comment by dougxc** [comment]
The Travis gate issue should now be [fixed](https://github.com/graalvm/graal-core/commit/9876e7672afe2eec7f32c889b6fcd71ec61ba8cd). Please rebase on top of the master branch now.

**Comment by theRealAph** [comment]
On 24/11/16 18:15, Roland Schatz wrote:
> rschatz commented on this pull request.
> 
> Thanks for the contribution!
> 
> I looked over the changes, they look good. I'm not really qualified to comment on technical details of Aarch64, but I guess you're testing that locally, so if it works it's ok ;)
> 
> This PR is currently failing the gate, but that is not your fault, the build was broken by a recent update to JDK 8u111. @dougxc is currently fixing that (https://github.com/graalvm/graal-core/pull/226), once that's in, can you please rebase on top of that, so we get a gate run?
> 
>>          if (fromBits == toBits) {
>              return inputVal;
> +        } else if (isJavaConstant(inputVal)) {
> 
> Is this code really alive? I find it weird that we're still seeing opportunities for folding constants during LIR generation...

It's called from emitCompare.  It's really a hangover from sub-word comparisons comparisons, which were removed by your earlier patch https://github.com/graalvm/graal-core/commit/e276a1dc73876559f2ada69797dcb961076b51f0. I could talke it out.

>> @@ -596,6 +624,8 @@ public void neg(int size, Register dst, Register src) {
>       */
>      @Override
>      public void add(int size, Register dst, Register src, int immediate) {
> +        if ((immediate & 7) != 0)
> 
> Please always use `{ }` even for one-liners.

Righto.

Andrew.

**Comment by theRealAph** [comment]
I pushed a small change to the branch and rebased it from master, but
now when I go to Compare changes between master and
aarch64_graal_misc_fixes it tells me that

"There isn’t anything to compare."

Username for 'https://github.com': theRealAph
Password for 'https://theRealAph@github.com':
Counting objects: 59, done.
Delta compression using up to 48 threads.
Compressing objects: 100% (5/5), done.
Writing objects: 100% (11/11), 686 bytes | 0 bytes/s, done.
Total 11 (delta 4), reused 0 (delta 0)
remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
To https://github.com/theRealAph/graal-core
   6f53bb5..d411bd8  aarch64_graal_misc_fixes -> aarch64_graal_misc_fixes
   39d44b6..03d3f8c  master -> master

Seemed to work OK.

**Comment by dougxc** [comment]
This PR stills shows 12 files changed so I think everything's fine. I see that there's a number of compilation errors and style issues as shown by Travis CI. Running the gate commands from `.travis.yml` locally should help fix these issues:
```
mx -v --strict-compliance gate --strict-mode --tags style,fullbuild
mx -v --strict-compliance gate --strict-mode --tags build,test
mx -v --strict-compliance gate --strict-mode --tags build,bootstraplite
```
Note that the `build` tag implies a clean so once one build successfully completes, you can omit that tag to run the other gate tasks.

**Comment by rschatz** [comment]
Github automatically updates the pull request when you push new commits to the same branch. That's why it says "nothing to compare", because the changes are already here.

File: graal/com.oracle.graal.asm.aarch64/src/com/oracle/graal/asm/aarch64/AArch64Assembler.java [hunk 2]
```diff
@@ -1786,7 +1786,7 @@ public void clz(int size, Register dst, Register src) {
      * @param dst general purpose register. May not be null, zero-register or the stackpointer.
      * @param src source register. May not be null, zero-register or the stackpointer.
      */
-    protected void rbit(int size, Register dst, Register src) {
+    public void rbit(int size, Register dst, Register src) {
```
Comment (sanzinger) [inline graal/com.oracle.graal.asm.aarch64/src/com/oracle/graal/asm/aarch64/AArch64Assembler.java]: If this method is going to be public, `TestProtectedAssembler` can be updated as well: https://github.com/graalvm/graal-core/blob/master/graal/com.oracle.graal.asm.aarch64.test/src/com/oracle/graal/asm/aarch64/test/TestProtectedAssembler.java#L339

**Comment by sanzinger** [comment]
Following classes are missing in this PR:
* `com.oracle.graal.lir.aarch64.AArch64Unary`
* `com.oracle.graal.lir.aarch64.AArch64ByteSwapOp`

**Comment by theRealAph** [comment]
On 25/11/16 10:37, Douglas Simon wrote:
> This PR stills shows 12 files changed so I think everything's fine. I see that there's a number of compilation errors and style issues as shown by Travis CI. Running the gate commands from `.travis.yml` locally should help fix these issues:
> ```
> mx -v --strict-compliance gate --strict-mode --tags style,fullbuild
> mx -v --strict-compliance gate --strict-mode --tags build,test
> mx -v --strict-compliance gate --strict-mode --tags build,bootstraplite
> ```
> Note that the `build` tag implies a clean so once one build successfully completes, you can omit that tag to run the other gate tasks.

AArch64Move.java:437: Each variable declaration must be in its own statement.

        try (ScratchRegister r1 = masm.getScratchRegister();
                        ScratchRegister r2 = masm.getScratchRegister()) {

Is this for real?

Andrew.

**Comment by dougxc** [comment]
> AArch64Move.java:437: Each variable declaration must be in its own statement.
> 
>         try (ScratchRegister r1 = masm.getScratchRegister();
>                         ScratchRegister r2 = masm.getScratchRegister()) {
> 
> Is this for real?
> 

Is that a Checkstyle warning? Something must be configured wrong as we have multiple var decls in try-with-resources in many, many places. Can you please link to the Travis output line showing this warning.

**Comment by theRealAph** [comment]
On 25/11/16 21:36, Douglas Simon wrote:
> Is that a Checkstyle warning? Something must be configured wrong as we have multiple var decls in try-with-resources in many, many places. Can you please link to the Travis output line showing this warning.

I think that I may have misinterpreted the warning.  Sorry for the noise.

**Comment by theRealAph** [comment]
On 25/11/16 10:37, Douglas Simon wrote:
> mx -v --strict-compliance gate --strict-mode --tags style,fullbuild

I'm getting a failure I don't understand.

Traceback (most recent call last):
  File "/home/graal/mx/mx_gate.py", line 278, in gate
    _warn_or_abort('Pylint not configured correctly. Cannot execute Pylint task.', args.strict_mode)
  File "/home/graal/mx/mx_gate.py", line 207, in _warn_or_abort
    reporter(msg)

It seems to be running pylint with the argument --primary.  But pylint
doesn't support the --primary argument.

Is some special version of pylint required?

**Comment by dougxc** [comment]
What happens when you run just `mx pylint`? It should check the version of pylint used:
https://github.com/graalvm/mx/blob/master/mx.py#L9978-L9990

The `--primary` argument is to the `mx pylint` command:
```
dsimon@kurz-3 ~/h/graal-core> mx pylint -h
usage: mx pylint [-h] [--primary] [--walk]

optional arguments:
  -h, --help  show this help message and exit
  --primary   limit checks to primary suite
  --walk      use tree walk find .py files
```
and should not be passed to the actual pylint execution.

**Comment by theRealAph** [comment]
On 26/11/16 12:19, Douglas Simon wrote:
> What happens when you run just `mx pylint`? It should check the version of pylint used:

mercury:graal $ mx pylint
require pylint version = 1.1.x (got 2.0.0)

Aha!  Ok, is there a download of pylint 1.1?  I can't build 1.1
from source because I get lots of syntax errors.

**Comment by dougxc** [comment]
You need to install pip (and it's dependency `astroid`) via [pip](https://pip.pypa.io/en/latest/installing/):
```
pip install astroid==1.1.0
pip install pylint==1.1.0
```

File: graal/com.oracle.graal.hotspot.aarch64/src/com/oracle/graal/hotspot/aarch64/AArch64HotSpotJumpToExceptionHandlerInCallerOp.java [hunk 3]
```diff
@@ -68,16 +68,19 @@ public AArch64HotSpotJumpToExceptionHandlerInCallerOp(AllocatableValue handlerIn
...
-        // Restore sp from fp if the exception PC is a method handle call site.
-        try (ScratchRegister sc = masm.getScratchRegister()) {
-            Register scratch = sc.getRegister();
-            AArch64Address address = masm.makeAddress(thread, isMethodHandleReturnOffset, scratch, 4, /* allowOverwrite */false);
-            masm.ldr(32, scratch, address);
-            Label noRestore = new Label();
-            masm.cbz(32, scratch, noRestore);
-            masm.mov(64, sp, fp);
-            masm.bind(noRestore);
+        if (System.getProperty("java.specification.version").compareTo("1.8") < 0) {
```
Comment (woess) [inline graal/com.oracle.graal.hotspot.aarch64/src/com/oracle/graal/hotspot/aarch64/AArch64HotSpotJumpToExceptionHandlerInCallerOp.java]: This check could use `Util.JAVA_SPECIFICATION_VERSION`. Although Graal does not even support versions < 1.8.

**Comment by gilles-duboscq** [comment]
There's something strange with the pull request, it looks like a lot of changes that are already in the repo have been duplicated. Some unfortunate rebase maybe?

**Comment by theRealAph** [comment]
On 12/12/16 19:06, Gilles Duboscq wrote:
> There's something strange with the pull request, it looks like a lot of changes that are already in the repo have been duplicated. Some unfortunate rebase maybe?

That's how I thought it was supposed to work: the branch with the pull
request should be based on the code it's supposed to be pulled into.
At least, that's how it works on the systems I normally use.

OK, so what am I supposed to do, then?

**Comment by gilles-duboscq** [comment]
> That's how I thought it was supposed to work: the branch with the pull
request should be based on the code it's supposed to be pulled into.

But it should only contain new changes, not copies of changes that are already in the repo. I'm not really sure what's going on, i'll have a look

**Comment by gilles-duboscq** [comment]
I pushed a merge with the master branch and now things look sane, i'm not quite sure what happened.
Looks good now

**Comment by theRealAph** [comment]
Where did you push this to?

**Comment by gilles-duboscq** [comment]
To this pull request (see the "Merge branch 'master' into aarch64_graal_misc_fixes" commit). I did it from the github UI though. I seems to have added the commit to your clone of the repo (and as a result to this PR).

**Comment by gilles-duboscq** [comment]
Andrew, do you mind if I rewrite your name to "Andrew Haley" in your commits where you used "theRealAph"?
I would have to rewrite the history though, so if you have already made commits based on those in this PR it might be better if you do that change yourself to avoid merge problems later.
If that's the case, what i did to rewrite locally was (while being on the branch with your commits):
```sh
git filter-branch --env-filter '
if [ "$GIT_COMMITTER_NAME" == "theRealAph" ] ; then
    export GIT_COMMITTER_NAME="Andrew Haley"
fi
if [ "$GIT_AUTHOR_NAME" == "theRealAph" ] ; then
    export GIT_AUTHOR_NAME="Andrew Haley"
fi
' $(git merge-base master HEAD)..HEAD
```

**Comment by theRealAph** [comment]
On 15/12/16 15:22, Gilles Duboscq wrote:
> Andrew, do you mind if I rewrite your name to "Andrew Haley" in your commits where you used "theRealAph"?

Absolutely not!  I chose "theRealAph" when I first signed up to GitHub
and was desperate to find a unique name.  I tried a ton of other
names and they were all taken.

Thanks,

Andrew.

**Comment by gilles-duboscq** [comment]
OK, i pushed the modified changesets in [my fork](https://github.com/gilles-duboscq/graal-core/tree/aarch64_graal_misc_fixes).
If you want to do some new changes, you can start working on a new branch based on my branch for example with:
```
$ git fetch git@github.com:gilles-duboscq/graal-core.git aarch64_graal_misc_fixes:gd-aarch64_graal_misc_fixes
$ git checkout -b my_new_aarch64_changes gd-aarch64_graal_misc_fixes
```

**Comment by gilles-duboscq** [comment]
The commits with the updated author names has been merged through 45e1839.

**Pull request closed without merging**